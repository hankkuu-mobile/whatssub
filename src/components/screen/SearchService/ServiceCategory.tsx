import * as React from 'react';

import { getString } from '../../../../STRINGS';
import styled from 'styled-components/native';

const Container = styled.View`
  background-color: ${({ theme }) => (theme.debug ? '#ddd' : 'transparent')};
  margin-left: 20;
  margin-right: 20;
  flex-direction: row;
  justify-content: flex-start;
  flex-wrap: wrap;
`;

const HorizontalScrollView = styled.ScrollView`
  padding-top: 12;
  padding-bottom: 12;
`;

const Category = styled.TouchableOpacity<{ selected?: boolean }>`
  background-color: ${({ theme, selected }) =>
    theme.debug ? 'red' : selected ? '#3e7eff' : 'transparent'};
  border-width: 1;
  border-color: #e1e4e7;
  border-radius: 17;
  padding-top: 9;
  padding-bottom: 9;
  padding-left: 15;
  padding-right: 15;
  margin-right: 5;
  justify-content: center;
  align-items: center;
`;

const Label = styled.Text<{ selected?: boolean }>`
  color: ${({ selected }) => (selected ? 'white' : '#a4a5a6')};
  font-size: 12;
  letter-spacing: -0.5;
`;

export interface SearchInputProps {
  categoryList: string[];
  initialCategory?: string;
  onSelect?: (category: string) => void;
  style?: any;
}

// TODO: error handle when there is no props.categoryList exists
function ServiceCategory(props: SearchInputProps) {
  const [selected, select] = React.useState(
    props.initialCategory ? props.initialCategory : '',
  );

  // TODO: Add assertion that category is unique. as it is being used as key
  return (
    <Container style={props.style}>
      <HorizontalScrollView horizontal={true}>
        {props.categoryList.map((category) => (
          <Category
            testID={`CATEGORY_SELECT_${category}`}
            key={category}
            selected={category === selected}
            onPress={() => {
              select(category);
              props.onSelect && props.onSelect(category);
            }}
          >
            <Label selected={category === selected}>
              {getString(category)}
            </Label>
          </Category>
        ))}
      </HorizontalScrollView>
    </Container>
  );
}

export default ServiceCategory;
