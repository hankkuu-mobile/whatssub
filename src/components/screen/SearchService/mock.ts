const THUMBNAIL =
  'https://facebook.github.io/react-native/docs/assets/favicon.png';

export interface Service {
  id: string;
  name: string;
  category: string;
  thumbnail: string;
  initialGroup: string;
}

// category key is already localized
export const SERVICE_LIST: Service[] = [
  {
    id: '1',
    name: '휴대폰ㆍ인터넷ㆍTV',
    category: 'CATEGORY_COMMUNICATION_EXPENSE',
    thumbnail: THUMBNAIL,
    initialGroup: '0',
  },
  {
    id: '2',
    name: '대출ㆍ보험ㆍ적금ㆍ예금ㆍ곗돈',
    category: 'CATEGORY_FINANCE',
    thumbnail: THUMBNAIL,
    initialGroup: '0',
  },
  {
    id: '3',
    name: '공과금ㆍ렌탈',
    category: 'CATEGORY_RESIDENCE_AND_MAINTENANCE',
    thumbnail: THUMBNAIL,
    initialGroup: '0',
  },
  {
    id: '4',
    name: '학원ㆍ기부ㆍ헌금',
    category: 'CATEGORY_ETC',
    thumbnail: THUMBNAIL,
    initialGroup: '0',
  },
  {
    id: '5',
    name: '넷플릭스',
    category: 'CATEGORY_ENTERTAINMENT',
    thumbnail: THUMBNAIL,
    initialGroup: '1',
  },
  {
    id: '6',
    name: '에버노트',
    category: 'CATEGORY_PRODUCTIVITY',
    thumbnail: THUMBNAIL,
    initialGroup: '1',
  },
  {
    id: '7',
    name: '폴라리스 오피스',
    category: 'CATEGORY_PRODUCTIVITY',
    thumbnail: THUMBNAIL,
    initialGroup: '1',
  },
  {
    id: '8',
    name: '유투브 프리미엄',
    category: 'CATEGORY_ENTERTAINMENT',
    thumbnail: THUMBNAIL,
    initialGroup: '1',
  },
  {
    id: '9',
    name: '멜론',
    category: 'CATEGORY_MUSIC',
    thumbnail: THUMBNAIL,
    initialGroup: '1',
  },
];

export const EXCEPTIONAL_CATEGORY_CASE = 'CATEGORY_RECOMMENDED';
export const CATEGORY_LIST = [
  'CATEGORY_COMMUNICATION_EXPENSE',
  'CATEGORY_FINANCE',
  'CATEGORY_RESIDENCE_AND_MAINTENANCE',
  'CATEGORY_ETC',
  'CATEGORY_ENTERTAINMENT',
  'CATEGORY_PRODUCTIVITY',
  'CATEGORY_MUSIC',
];
// push the exceptional case to the first index
CATEGORY_LIST.unshift(EXCEPTIONAL_CATEGORY_CASE);

export interface SeviceGroupTable {
  [key: string]: { key: string; order: number };
}
// title order lookup table
// `order` should starts from 0 and consecutively
// `order` is being used when initial case
export const INITIAL_GROUP_ORDER_TABLE: SeviceGroupTable = {
  0: { key: 'SERVICE_GROUP_LIVING_EXPENSE', order: 0 },
  1: { key: 'SERVICE_GROUP_SUBSCRIPTION_SERVICE', order: 1 },
};
