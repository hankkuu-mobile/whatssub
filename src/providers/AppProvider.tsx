import * as Font from 'expo-font';

import React, { useEffect, useReducer, useState } from 'react';

import { AsyncStorage } from 'react-native';
import { ThemeType } from '../types';

interface Context {
  state: State;
  changeTheme: (theme?: ThemeType) => void;
}
const AppContext = React.createContext<Context>(null);
const AppConsumer = AppContext.Consumer;

interface ChangeThemeTypeAction {
  type: 'change-theme-type-mode';
  payload: {
    theme?: ThemeType;
  };
}

interface Props {
  theme?: ThemeType;
  children?: any;
  doNotWaitFont?: boolean;
}

interface State {
  theme?: ThemeType;
}

export const initialState: State = {
  theme: ThemeType.LIGHT,
};

const reducer = (state: State, action: ChangeThemeTypeAction) => {
  switch (action.type) {
    case 'change-theme-type-mode':
      // if payload.theme doesn't exist, toggle
      if (!action.payload.theme) {
        return {
          ...state,
          theme:
            state.theme === ThemeType.LIGHT
              ? ThemeType.DARK
              : ThemeType.LIGHT,
        };
      }
      // else change it with payload.theme
      return {
        ...state,
        theme: action.payload.theme,
      };
    default:
      return state;
  }
};

function AppProvider(props: Props) {
  const [state, dispatch] = useReducer(reducer, initialState);
  const [fontLoading, setFontLoading] = useState(true);

  // by encapsulating dispatch, it is easier to use (such as passing through screenProps)
  const changeTheme: (theme?: ThemeType) => void = (theme) => {
    dispatch({
      type: 'change-theme-type-mode',
      payload: {
        theme,
      },
    });
  };

  // initialize theme
  useEffect(() => {
    changeTheme(props.theme ? props.theme : initialState.theme);
  }, [props.theme]);

  // Load font and then use saved theme from local storage
  const getCurrentThemeType = async () => {
    const value = (await AsyncStorage.getItem('theme')) as ThemeType;
    value && changeTheme(value);
  };
  const loadFonts = async () => {
    await Font.loadAsync({
      'spoqa-han-sans-bold': require('../../assets/fonts/SpoqaHanSans/SpoqaHanSans-Bold.ttf'),
      'spoqa-han-sans-regular': require('../../assets/fonts/SpoqaHanSans/SpoqaHanSans-Regular.ttf'),
    });

    setFontLoading(false);
  };
  useEffect(() => {
    loadFonts();
    getCurrentThemeType();
  }, []);

  useEffect(() => {
    AsyncStorage.setItem('theme', state.theme);
  }, [state.theme]);

  return (
    <AppContext.Provider
      value={{
        state,
        changeTheme,
      }}
    >
      {fontLoading && !props.doNotWaitFont ? null : props.children}
    </AppContext.Provider>
  );
}

export { AppConsumer, AppProvider, AppContext };
