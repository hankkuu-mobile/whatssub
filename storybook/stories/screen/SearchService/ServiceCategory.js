import { CATEGORY_LIST } from '../../../../src/components/screen/SearchService/mock';
import React from 'react';
import ServiceCategory from '../../../../src/components/screen/SearchService/ServiceCategory';
import { storiesOf } from '@storybook/react-native';
import styled from 'styled-components/native';

const Container = styled.SafeAreaView`
  flex: 1;
`;

const SimpleContainerDecorator = (storyFn) => (
  <Container>{storyFn()}</Container>
);

storiesOf('SearchService', module)
  .addDecorator(SimpleContainerDecorator)
  .add('â””ServiceCategory', () => (
    <ServiceCategory
      categoryList={CATEGORY_LIST}
      onSelect={(category) => {
        // eslint-disable-next-line no-console
        console.log('category selected', category);
      }}
      initialCategory={CATEGORY_LIST[0]}
    />
  ));
