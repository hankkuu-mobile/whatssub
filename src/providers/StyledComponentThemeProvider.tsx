import React, { useContext, useState } from 'react';

import { AppContext } from './AppProvider';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'styled-components';
import { createTheme } from '../theme';

const StyledComponentThemeProvider = (props) => {
  const { state } = useContext(AppContext);
  const [debug, setDebug] = useState(false);
  return (
    <ThemeProvider theme={{ ...createTheme(state.theme), debug, setDebug }}>
      {props.children}
    </ThemeProvider>
  );
};

StyledComponentThemeProvider.propTypes = {
  children: PropTypes.node,
};

export default StyledComponentThemeProvider;
