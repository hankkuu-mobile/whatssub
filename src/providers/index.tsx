import { AppConsumer, AppContext, AppProvider } from './AppProvider';

import PropTypes from 'prop-types';
import React from 'react';
import StyledComponentThemeProvider from './StyledComponentThemeProvider';
import { ThemeType } from '../types';

interface Props {
  theme?: ThemeType;
  children?: any;
  doNotWaitFont?: boolean;
}

const InitialProviders = ({ children, ...rest }: Props) => (
  <AppProvider {...rest}>
    <StyledComponentThemeProvider>{children}</StyledComponentThemeProvider>
  </AppProvider>
);

InitialProviders.propTypes = {
  children: PropTypes.node,
};

export { AppProvider, AppConsumer, AppContext, InitialProviders };
