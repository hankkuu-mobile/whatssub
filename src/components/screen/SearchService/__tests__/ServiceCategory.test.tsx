import 'react-native';

import * as React from 'react';

import { fireEvent, render, wait } from '@testing-library/react-native';
import renderer, { act } from 'react-test-renderer';

import { CATEGORY_LIST } from '../mock';
import { InitialProviders } from '../../../../providers';
import ServiceCategory from '../ServiceCategory';
import i18n from 'i18n-js';

i18n.locale = 'ko';
let component;
let props: any;

const createTestProps = (props: Object) => ({
  onSelect: jest.fn(),
  categoryList: CATEGORY_LIST,
  initialCategory: CATEGORY_LIST[0],
});

const setup = () => {
  props = createTestProps({});
  component = (
    <InitialProviders doNotWaitFont>
      <ServiceCategory {...props} />
    </InitialProviders>
  );
};

describe('[ServiceCategory] render', () => {
  beforeAll(setup);
  it('component and snapshot matches', () => {
    const rendered: renderer.ReactTestRendererJSON = renderer
      .create(component)
      .toJSON();
    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });
});

it('should call onSelect(category name) when relevant category is touched', () => {
  const category = CATEGORY_LIST[0];
  const { getByTestId } = render(component);
  const ComponentsToTest = getByTestId(`CATEGORY_SELECT_${category}`);
  act(() => {
    fireEvent.press(ComponentsToTest);
  });
  expect(props.onSelect).toBeCalledWith(category);
});
