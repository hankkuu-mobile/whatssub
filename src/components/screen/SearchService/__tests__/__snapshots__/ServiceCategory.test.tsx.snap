// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[ServiceCategory] render component and snapshot matches 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
        "flexDirection": "row",
        "flexWrap": "wrap",
        "justifyContent": "flex-start",
        "marginLeft": 20,
        "marginRight": 20,
      },
    ]
  }
>
  <RCTScrollView
    horizontal={true}
    style={
      Array [
        Object {
          "paddingBottom": 12,
          "paddingTop": 12,
        },
      ]
    }
  >
    <View>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#3e7eff",
            "borderColor": "#e1e4e7",
            "borderRadius": 17,
            "borderWidth": 1,
            "justifyContent": "center",
            "marginRight": 5,
            "opacity": 1,
            "paddingBottom": 9,
            "paddingLeft": 15,
            "paddingRight": 15,
            "paddingTop": 9,
          }
        }
        testID="CATEGORY_SELECT_CATEGORY_RECOMMENDED"
      >
        <Text
          selected={true}
          style={
            Array [
              Object {
                "color": "white",
                "fontSize": 12,
                "letterSpacing": -0.5,
              },
            ]
          }
        >
          추천
        </Text>
      </View>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderColor": "#e1e4e7",
            "borderRadius": 17,
            "borderWidth": 1,
            "justifyContent": "center",
            "marginRight": 5,
            "opacity": 1,
            "paddingBottom": 9,
            "paddingLeft": 15,
            "paddingRight": 15,
            "paddingTop": 9,
          }
        }
        testID="CATEGORY_SELECT_CATEGORY_COMMUNICATION_EXPENSE"
      >
        <Text
          selected={false}
          style={
            Array [
              Object {
                "color": "#a4a5a6",
                "fontSize": 12,
                "letterSpacing": -0.5,
              },
            ]
          }
        >
          통신비
        </Text>
      </View>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderColor": "#e1e4e7",
            "borderRadius": 17,
            "borderWidth": 1,
            "justifyContent": "center",
            "marginRight": 5,
            "opacity": 1,
            "paddingBottom": 9,
            "paddingLeft": 15,
            "paddingRight": 15,
            "paddingTop": 9,
          }
        }
        testID="CATEGORY_SELECT_CATEGORY_FINANCE"
      >
        <Text
          selected={false}
          style={
            Array [
              Object {
                "color": "#a4a5a6",
                "fontSize": 12,
                "letterSpacing": -0.5,
              },
            ]
          }
        >
          금융
        </Text>
      </View>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderColor": "#e1e4e7",
            "borderRadius": 17,
            "borderWidth": 1,
            "justifyContent": "center",
            "marginRight": 5,
            "opacity": 1,
            "paddingBottom": 9,
            "paddingLeft": 15,
            "paddingRight": 15,
            "paddingTop": 9,
          }
        }
        testID="CATEGORY_SELECT_CATEGORY_RESIDENCE_AND_MAINTENANCE"
      >
        <Text
          selected={false}
          style={
            Array [
              Object {
                "color": "#a4a5a6",
                "fontSize": 12,
                "letterSpacing": -0.5,
              },
            ]
          }
        >
          주거/관리
        </Text>
      </View>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderColor": "#e1e4e7",
            "borderRadius": 17,
            "borderWidth": 1,
            "justifyContent": "center",
            "marginRight": 5,
            "opacity": 1,
            "paddingBottom": 9,
            "paddingLeft": 15,
            "paddingRight": 15,
            "paddingTop": 9,
          }
        }
        testID="CATEGORY_SELECT_CATEGORY_ETC"
      >
        <Text
          selected={false}
          style={
            Array [
              Object {
                "color": "#a4a5a6",
                "fontSize": 12,
                "letterSpacing": -0.5,
              },
            ]
          }
        >
          기타
        </Text>
      </View>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderColor": "#e1e4e7",
            "borderRadius": 17,
            "borderWidth": 1,
            "justifyContent": "center",
            "marginRight": 5,
            "opacity": 1,
            "paddingBottom": 9,
            "paddingLeft": 15,
            "paddingRight": 15,
            "paddingTop": 9,
          }
        }
        testID="CATEGORY_SELECT_CATEGORY_ENTERTAINMENT"
      >
        <Text
          selected={false}
          style={
            Array [
              Object {
                "color": "#a4a5a6",
                "fontSize": 12,
                "letterSpacing": -0.5,
              },
            ]
          }
        >
          엔터테인먼트
        </Text>
      </View>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderColor": "#e1e4e7",
            "borderRadius": 17,
            "borderWidth": 1,
            "justifyContent": "center",
            "marginRight": 5,
            "opacity": 1,
            "paddingBottom": 9,
            "paddingLeft": 15,
            "paddingRight": 15,
            "paddingTop": 9,
          }
        }
        testID="CATEGORY_SELECT_CATEGORY_PRODUCTIVITY"
      >
        <Text
          selected={false}
          style={
            Array [
              Object {
                "color": "#a4a5a6",
                "fontSize": 12,
                "letterSpacing": -0.5,
              },
            ]
          }
        >
          생산성
        </Text>
      </View>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderColor": "#e1e4e7",
            "borderRadius": 17,
            "borderWidth": 1,
            "justifyContent": "center",
            "marginRight": 5,
            "opacity": 1,
            "paddingBottom": 9,
            "paddingLeft": 15,
            "paddingRight": 15,
            "paddingTop": 9,
          }
        }
        testID="CATEGORY_SELECT_CATEGORY_MUSIC"
      >
        <Text
          selected={false}
          style={
            Array [
              Object {
                "color": "#a4a5a6",
                "fontSize": 12,
                "letterSpacing": -0.5,
              },
            ]
          }
        >
          음악
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
